Add locked lesson free course notices

renatho opened the pull request:
Part of #4445

### Changes proposed in this Pull Request

* It introduces notices  (logged out and not enrolled students) in the lesson pages for free courses.
* It also contains some redirects logic if the user is logged out.
  * If they click on "Sign in", they log in, then they will be redirected to the same page they were.
  * If they click on "Take course", they log in, then they will be redirected to the course page, with the button "Start course" instead of "Take course" (or any other name they used).
* Notice that the redirects after login may not work if users are using a custom login/registration page. It's was implemented using `redirect_to` query string, which is the same used by WP core login.
* It also refactors some code to use the new `sensei_user_registration_url` function.

### Testing instructions

<!--
If the functionality of this PR is behind a feature flag, please include the
following testing step, using the correct name for the feature flag. (If you
aren't sure, just ignore this step)

* Enable the feature flag: `add_filter( 'sensei_feature_flag_{THE_FLAG_NAME}', '__return_true' );`
-->
* Enable the feature flag: `add_filter( 'sensei_feature_flag_course_theme', '__return_true' );`
* Create a course.
* Add at least one lesson.
* Select the "Sensei Theme" as the "Course Theme" (sidebar).
* Logged out, access a lesson page, and make sure you see a notice with "Take course" and "Sign in" buttons.
* Using the Sign in button, make sure you're redirected back to the lesson page you were.
* Using the Take course button, make sure you're redirected to the course with a button "Start course" instead of "Take course" (just the label was replaced for that redirect case).
* Now, logged in as a student, try to access a lesson page, and make sure you see a notice with the "Take course" button only.
* Make sure that by clicking on that button, you get enrolled in the course. When https://github.com/Automattic/sensei/pull/4496 is completed, the user will be redirected to the first lesson after the enrollment.

<!-- Add the following only if there are new/updated actions or filters. Please provide a brief description of what they do and any arguments they may take. Be sure to also add the "Hooks" label to this PR. -->
### New/Updated Hooks

* `sensei_login_success_redirect_url` - The old `sesei_login_success_redirect_url` renamed.

<!-- Add the following only if there is any code that is being deprecated. Please list the replacement function or hook that should be called instead, if applicable. Be sure to also add the "Deprecation" label to this PR. -->
### Deprecated Code

* `sesei_login_success_redirect_url` - Deprecated because the typo in the filter name.

<!--
Helpful tips for screenshots:
https://en.support.wordpress.com/make-a-screenshot/
-->
### Screenshot / Video

https://user-images.githubusercontent.com/876340/145272704-40b977d8-71e1-4011-82f9-6439125be868.mov

On 0:13, notice that it will redirect to the first lesson after https://github.com/Automattic/sensei/pull/4496 is completed.

#### Another case, in a course with course prerequisite.

<img width="940" alt="Screen Shot 2021-12-08 at 16 27 08" src="https://user-images.githubusercontent.com/876340/145272783-6159274e-9189-4823-b311-906819158e6b.png">

renatho commented on includes/course-theme/class-sensei-course-theme-lesson.php:
```diff
@@ -116,8 +116,14 @@ private function maybe_add_quiz_results_notice() {
 	 * Maybe add lesson prerequisite notice.
 	 */
 	private function maybe_add_prerequisite_notice() {
+		$lesson_id = \Sensei_Utils::get_current_lesson();
```
It fixes a case to not show the prerequisite notice if the user is not enrolled.

renatho commented on assets/css/sensei-course-theme/base.scss:
```diff
@@ -1,9 +1,10 @@
-.sensei-course-theme__frame {
-
+.sensei-course-theme {
```
Moved the variables to a level up, for now, so we can use it in other parts, like the notices.

renatho commented on includes/blocks/class-sensei-block-take-course.php:
```diff
@@ -55,6 +55,16 @@ public function render_take_course_block( $attributes, $content ): string {
 				Sensei()->notices->add_notice( Sensei()->course::get_course_prerequisite_message( $course_id ), 'info', 'sensei-take-course-prerequisite' );
 				$html = $this->render_disabled( $content );
 			} else {
+				// Replace button label in case it's coming from a sign in with redirect to take course.
```
It will be expanded in the future as part of https://github.com/Automattic/sensei/issues/4493

dadish commented on includes/course-theme/class-sensei-course-theme-lesson.php:
```diff
@@ -142,4 +149,73 @@ private function maybe_add_prerequisite_notice() {
+				$notices->add_notice(
+					'locked_lesson',
+					Sensei()->course::get_course_prerequisite_message( $course_id ),
+					__( 'You don\'t have access to this lesson', 'sensei-lms' ),
```
Should this be "You don't have access to this course" instead?

renatho replied:
I think it should be "lesson" to keep consistent with the other messages because the other ones also talk about course access. cc @sruj09 to confirm. Here we're talking about:

<img width="940" alt="Screen Shot 2021-12-08 at 16 27 08" src="https://user-images.githubusercontent.com/876340/145272783-6159274e-9189-4823-b311-906819158e6b.png">

Also, good catch about the redirect issue. I'll explore a little because when using WP registration it worked for admin too. Probably it's something in our login.

sruj09 replied:
Confirming that it should be "lesson". :)
Can we tweak the pre-requisite message as well? "You must complete the prerequisites to view this lesson"

renatho replied:
Hey Srujana!

We can tweak this. Just before doing that, let me give you some context here to see what you think.

First, just to confirm that there's no confusion with the other prerequisite message, it's related to the Course prerequisite (the lesson is another message), okay? :)

The message we're using here is the same message being used on the Course page (with the blocks, and even before the blocks - in the "legacy version"). And this message is also associated with a filter, where the users can customize it.

I see 2 options we can go here:

1) Update the message that is being used on the Course page too.
2) Just ignore that, and use a custom message here (probably okay since it's a different scope - the lesson page).

dadish replied:
Got it. Yeah, the message sounds about right.

renatho replied:
I'm merging this PR, but we can continue the discussion and tweak it in a separate PR. ;)

dadish reviewed (dismissed):
Cool beanz! One thing I noticed is if logging in as an admin, the redirections does not work. It logs into the WP admin. I suspect this is how WordPress works. Should not be an issue until the teacher is testing out his setup and it does not work for him, because his account is an admin.

renatho commented on includes/course-theme/class-sensei-course-theme-lesson.php:
```diff
@@ -142,4 +149,79 @@ private function maybe_add_prerequisite_notice() {
+
+			// Take course URL.
+			$take_course_redirect_url = add_query_arg( 'take_course_sign_in', '1', get_permalink( $course_id ) );
+			$take_course_url          = $user_can_register ? sensei_user_registration_url( true, $take_course_redirect_url ) : sensei_user_login_url( $take_course_redirect_url );
```
As the default option, we're sending logged-out users to the registration form. If registration is not activated, we send them to the login form.

renatho commented:
@dadish , Thank you for your review! About your catch, I realized we had a filter with redirect. That filter also used a `redirect_to`, but only from post, so I fixed it here: https://github.com/Automattic/sensei/pull/4515/commits/d2dc4ee2bf70a75f5defd6ac958a37c3f7a19baa

Also, I did a small refactor here: https://github.com/Automattic/sensei/pull/4515/commits/ea688a452f470a5a2390eabdeff33dcabb1d3ae4 So we improve the parity between `sensei_user_registration_url` and `sensei_user_login_url`.

Also, in the same commit, I introduced a check if the registration is allowed. We redirect the logged-out user to the registration form by default, otherwise, we redirect them to the login form. ;)
